Added ProMPs and Kinematics functionalities

souljaboy764 opened the pull request:
Implemented Probabilistic Movement Primitives, specifically the conditioning aspect along with a general kinematics base class to perform forward and inverse kinematics. This needs to be extended for specific robots, providing the corresponding forward model. The code is adapted from https://github.com/sebasutp/promp

#### TODO:
- Add examples for using the ProMPs and conditioning
- Implement the EM-based learning for ProMPs as in https://github.com/sebasutp/promp

souljaboy764 committed:
Added Code for ProMPs and Kinematics

souljaboy764 committed:
Added Author information for Kinematics class

souljaboy764 committed:
Minor bug fix in kinematics.py

souljaboy764 committed:
Minor bug fix in ProMP and updated init

souljaboy764 commented:
Need to test a bit more before this can be merged

souljaboy764 committed:
Bug fixes in ProMP and Kinematics

souljaboy764 commented:
Tested the ProMP and Kinematics codes and they are working now.
Need to add examples to show how to use them.

joe-campbell commented:
Sorry for the late followup, somehow I missed when this was opened.

From your last comment, if you could add examples for how to use this so I could better understand the changes then that would be great! Something like a minimal tutorial script or such.

Commit:
Added ProMP example

Committer: souljaboy764

souljaboy764 commented:
I have added an example file at examples/promp_example.py and have also added a notebook about it as well.

souljaboy764 committed:
Changed to imports to be Python3 compatible

souljaboy764 commented:
I was having issues when importing the package in python3 due to the relative imports. I have changed the imports in the files to support python3 style syntax. I am now able to run the basic examples on python3.

hbaradius commented:
We should add souljaboy to the list of contributers to the IntPrim library. Very cool stuff!

souljaboy764 committed:
Added recursive conditioning example and fixed ProMP conditioning instalbility

joe-campbell commented:
Thanks for the latest push! Apologies for the lack of activity on my end, I was caught up in some other work. I will look at this in some detail over the weekend try to merge it. By the way, I noticed there are some hard coded paths in this latest commit, e.g.,

with open('/home/corobi/playground/intprim_test/strike_mov.npz','rb') as f:

Can this be changed such that any path is able to be set as a parameter? Otherwise other people won't be able to run these tests without modifying the source code first.

souljaboy764 commented:
Oh, I must have overlooked this.
I'll add the path to be given as a command line argument and add a comment on the usage.

souljaboy764 commented:
@joe-campbell Any suggestions on how to do that for the python notebook version of the example? Or should I just include the [`strike_mov.npz`](https://github.com/sebasutp/promp/blob/master/examples/strike_mov.npz) file from https://github.com/sebasutp/promp which I've used for the examples?

joe-campbell commented:
Hmm, there's no license given in that other library so I would prefer not to include any other data files directly. How about if we just provide an instruction in the notebook to download it and unzip it to a relative path location that the notebook looks for?

The notebook won't run without that, but I think it's ok. And the locally compiled version which is pushed up to github will still show what it looks like when that file is present.

souljaboy764 committed:
Updated examples

souljaboy764 commented:
I updated the examples and the notebook as well. I had also changed some of the code to be python3 compatible in one of the previous commits. It works on my system but would need to be tested further before claiming it to be properly python3 compatible.

souljaboy764 committed:
Merge branch 'ir-lab:master' into master

souljaboy764 committed:
added single sample spatial conditioning

souljaboy764 committed:
Added paper in README

souljaboy764 committed:
Bug fixes for EKF

souljaboy764 committed:
Minor changes

souljaboy764 commented:
I found some bugs in the EKF filtering which I (hopefully) fixed. I am currently adding single sample conditioning to the EKF as the previous commit was only for the KF.

souljaboy764 committed:
Minor changes
